Bf cleanup travis

unitware opened the pull request:
When troubleshooting https://github.com/betaflight/betaflight/issues/2708 I was looking in .travis.yml and travis.sh. and also in the build logs on travis. 

Does anyone use the functions publish_url, publish_docs, publish_meta and the target targets?

This PR has:
1. removed all but the $GOAL related stuff
2. made sure that we can use the current build goals {test, all} and also specify like {test, naze32, spracing, ..} - ie a selected number of targets for our work branches
3. made an attempt to bring down build times to approximately 7 minutes. This is done with a manual grouping of targets in the makefile and a few new targets that are invoked from the travis.yml file
```
> make help
...
test                 : run unit tests
targets-group-1      : build some targets        
targets-group-2      : build some targets       
targets-group-3      : build some targets       
targets-group-rest   : build the rest of the targets (not listed in group 1, 2 or 3)
...
```
The build is divided in five jobs to make full use of the travis account plan (which is five concurrent builds, at least for me)

Build times are now approximately 7 minutes for all targets including tests. Before it was about 23 minutes.

Some more work can be done by moving some targets to the job running tests but I decided against it for now - keeping the test separated from the target builds.

martinbudden commented:
Good stuff. Especially welcome is the reduction in build times.

unitware commented:
@wind0r wrote:
```
Wow. That is a really good speed up. Thank you very much and have a good night
But those groups should get a better names. Maybe group names and sorting  by architecture
```

I did sort by architecture at first, but the different build times were not equal so that gave ~10 minutes build time. The ordering is now "mostly F1 + some F7", "F3", "F4", the rest. I might have moved a few more targets but then I thought that it really didn't matter to much.

One thing we can do is to move "important" targets - i.e. the newest, in development and most buggy to the top of each group, along with the most constrained ones. That would give even faster feedback.

I'm totally open to suggestions of grouping, ordering and naming of the groups.

martinbudden commented:
In terms of putting certain targets at the top of the lists, I suggest COLIBRI_RACE, since that is the one that probably most frequently fails (due to forgetting to update BST stuff) and STM32F3DISCOVERY, since that probably has the most `USE_` build flags defined.

AndersHoglund commented:
This is good work! 
Sorting and categorizing the lists in one way or the other is futile and temporal. One day this target is "more important", next day that target.  And next week a new troublesome target will turn up in the "rest" list... Just sort them alphabetically. Hint: Use $(sort $(list)) in the same way and in the same place as $(VALID_TARGETS) are sorted. 

And for naming I think 1, 2, 3. etc is just as good as any, as we should not try any categories.  

At irregular intervals the lists need to be maintained and balanced too, especially  items "the reset" -list needs to be moved when it becomes too big. 

One thing I would like to see added are goals to list the target-groups. We have `make targets`, similar goals for the groups would be nice. Maybe as there own goals, or just printouts added to the existing `targets `goal.

AndersHoglund commented:
A few more comments:

Help texts needs some alignment. Try `make help` and you will see. Some trailing white space should also be removed.

Some re-arrangement of where in the Makefile the lists are defined is also needed. It is now inserted among the goals/rules.  Should be moved to a place right after where VALID_TARGETS and OFFICIAL_TARGETS are defined. Separation of data defs and rules.

There should also be a `TARGETS_REST_LIST` defined, same as all other lists. And the goal `targets-group-rest` depending on that list. 

There are also an unwritten Makefile style guide to have `all:` as the first default goal and rule. The goals  for all the group lists are added ahead of this. But this style is already broken and also overridden by the DEFAULT_GOAL definitions. No biggie as the Makefile is a bit of a mess already, but maybe we should try make it a bit more structured. 

There are also some tabs that needs to be cleaned out. 3 lines starting at 1129 have leading tabs. 

Remove `GROUP_4_LIST` that now is commented out. Or put it into work.

unitware commented:
Thanks for your feedback, I have reworked it a bit and now the machine running unittests are also building the rest, thus reducing the time a tiny bit more. Squashed the commits a bit.

Keep the feedback coming (or merge, I'm fairly satisfied with it as it stands).

AndersHoglund commented on Makefile:
```diff
@@ -110,6 +110,94 @@ VALID_TARGETS  := $(subst /,, $(subst ./src/main/target/,, $(VALID_TARGETS)))
+#  	NUCLEOF7 \
+#  	OMNIBUSF4 \
+
+GROUP_TEST_AND_REST_TARGETS := $(filter-out $(GROUP_1_TARGETS) $(GROUP_2_TARGETS) $(GROUP_3_TARGETS) $(GROUP_4_TARGETS), $(VALID_TARGETS))
```
List name not consistent with content. No tests in the list.

AndersHoglund commented on Makefile:
```diff
@@ -110,6 +110,94 @@ VALID_TARGETS  := $(subst /,, $(subst ./src/main/target/,, $(VALID_TARGETS)))
+# 	ZCOREF3 \
+#  	KROOZX \
+#  	NUCLEOF7 \
+#  	OMNIBUSF4 \
```
Please remove commented code.

AndersHoglund commented on Makefile:
```diff
@@ -1102,16 +1190,34 @@ $(OBJECT_DIR)/$(TARGET)/%.o: %.S
+
+## targets-group-test-and-rest :
+##                     build the rest of the targets (not listed in group 1, 2 or 3)
+targets-group-test-and-rest: test $(GROUP_TEST_AND_REST_TARGETS)
```
Tests and builds should be done as separate goals. Not mixing tests and (some) target builds.

martinbudden replied:
I agree. The test goal should just do tests, even if this results in slightly longer execution time.

unitware replied:
This is to utilize the build server fully, the test takes much shorter time and then that build slot is wasted unless we fill in with real builds.

AndersHoglund replied:
We should not mix two different things anyhow. When this job is in red, it is not obvious if the tests or some target build failed. Unless digging into the log. Split the two different types of jobs. If tests are short and quick, that slot will be free to take new jobs anyhow. Plus the extra overhead.
Structure and order above detailed hand optimization.

martinbudden replied:
I agree with Anders, we should not mix the two things. Yes, it produces a small improvement in build time, but it also saves developer time being immediately able to see if it is a build failure or a test failure.

AndersHoglund commented on Makefile:
```diff
@@ -1205,9 +1311,10 @@ help: Makefile make/tools.mk
+	$(V0) @echo "Valid targets:    $(VALID_TARGETS)"
+	$(V0) @echo "Target:           $(TARGET)"
+	$(V0) @echo "Base target:      $(BASE_TARGET)"
+	$(V0) @echo "Combined targets: targets-group-1, targets-group-2, targets-group-3, targets-group-4, targets-group-test-and-rest"
```
Why list the internal list names? The content of the lists are more interesting.

unitware replied:
it is a list of makefile targets, not what will happen when the targets are called. but I can add the contents instead, no problem

AndersHoglund replied:
Aha, OK. Add a double-hash comment above each goals, if the intent is to document the goals. As it is done on other goals. 
Listing the content of the lists are also important.

unitware commented:
ok, reverted the test and rest stuff, and adressed the other comments (unless I forgot anything)
We are using the free open source plan on travis right? or do we have more than five executing josts in parallell?

AndersHoglund commented:
OK, looks much better now. Yes, I think we only have 5 free slots maximum. Have not checked what other plans there are. Maybe some of the FC designers/producers/sellers here could sponsor a bigger plan....

Too bad there are 2 different things in one PR, target-groups and publish clean-up. Not sure I want to take responsibility to merge the later. The target-groups alone would have been easy. Even if betaflight does not use any of the publishing stuff, CF might do that. I really do not know. The removal here might cause later havoc when BF is merged to CF.  @martinbudden please decide to merge or not....... 

Lesson for the future, only one issue per PR. Never ever "oh by the way, I can fix this too...". But that is an easy trap to fall into. I do that all the time.....

unitware commented:
It's still two commits, so it can be separated - however if we're not uing it than it's not that hard for whomever to fix it when they miss it. I'm generally against stuff that is not actively used and that has rotten (like the original cause of me lifting on these rocks: https://github.com/betaflight/betaflight/issues/2708)

unitware commented:
There is not much documentation left here since it's in the wiki now?

unitware commented:
@hydra Are you using the travis "publish_docs" or "Publish_meta" features in .travis.yml?

hydra commented:
@unitware I plan to re-instate the manual and docs building process.  I've not had the chance to look at that yet but would prefer not to delete it.  Perhaps build a very short manual for bf that just has a logo and a link to a wiki page but build the docs from the same source directory in both projects?

hydra commented on .travis.sh:
```diff
@@ -1,66 +1,6 @@
-		curl -k "${CURL_BASEOPTS[@]}" "${CURL_PUB_BASEOPTS[@]}" --form "file=@${TARGET_FILE}" ${PUBLISH_URL} || true
-		exit 0;
-	fi
-elif [ $GOAL ] ; then
```
would prefer to keep the publish* goals working please so we can keep the makefile aligned in the forks.

unitware commented:
Ok, I reverted parts of the cleanup and fixed the travis TARGET=nn functionality, it was broken in a few ways. (fail did not register, target filename ad no version).

Also split in four groups again, the "rest" will run after the test is done. We're approaching a build time of six minutes.

AndersHoglund commented:
Please, only revert the publishing stuff. 
This PR should have the target-group builds only. Make separate PRs for the publishing and file name version stuff. Two or three totally different issues requires separate PRs, not mixed up in one.

Not sure I agree with target file name versioning and the format.... But that is another matter, in another PR.

unitware commented:
Closing this since I have splited the PR in two:
https://github.com/betaflight/betaflight/pull/2723 - Speedup build by grouping the builds in a few groups
https://github.com/betaflight/betaflight/pull/2724 - Fix issue #2708, travis did not fail if build TARGET=.. failed
